[ET-1609] Enhancement - Added new currency format options for Tickets Commerce tickets.

rafsuntaskin opened the pull request:
### 🎫 Ticket

[ET-1608]
[ET-1609]
[ET-1610]

### 🗒️ Description

<!-- Please describe what you have changed or added -->
<!-- What types of changes does your code introduce? -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Include any important information for reviewers -->
<!-- Etc, etc, etc -->

### 🎥 Artifacts <!-- if applicable-->
🎥 screencast(s): https://d.pr/v/z4L6AC

### ✔️ Checklist
- [x] I've included a changelog entry in the readme.txt file. <!-- Confirm that it includes the ticket ID. -->
- [x] My code is tested. <!-- Check that tests are passing and DO NOT merge if they're failing. -->
- [x] My code has [proper inline documentation](https://the-events-calendar.github.io/products-engineering/docs/code-standards/).


[ET-1608]: https://theeventscalendar.atlassian.net/browse/ET-1608?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ET-1609]: https://theeventscalendar.atlassian.net/browse/ET-1609?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ET-1610]: https://theeventscalendar.atlassian.net/browse/ET-1610?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

codingmusician commented on src/Tickets/Commerce/Hooks.php:
```diff
@@ -728,4 +731,36 @@ public function register_event_compatibility_hooks() {
+			return $options;
+		}
+		// if they are not same then update the rest format options from default map.
+		$currency_data = Base_Commerce\Utils\Currency::get_default_currency_map();
```
Should we add a check after this line to make sure that the currency data exists before using it?

rafsuntaskin replied:
Updated as per suggestion. Thanks 👍

codingmusician commented on src/Tickets/Commerce/Settings.php:
```diff
@@ -287,11 +316,36 @@ public function get_settings() {
 				'type'            => 'dropdown',
 				'label'           => esc_html__( 'Currency Code', 'event-tickets' ),
-				'tooltip'         => esc_html__( 'The currency that will be used for Tickets Commerce transactions.', 'event-tickets' ),
+				'tooltip'         => esc_html__( 'The currency that will be used for Tickets Commerce transactions.', 'event-tickets' ) . '<b> Format Preview : ' . $preview_amount . '</b>',
```
"Format Preview" needs to be translatable, too.

juanfra commented on src/Tickets/Commerce/Hooks.php:
```diff
@@ -728,4 +731,17 @@ public function register_event_compatibility_hooks() {
+	 * @return array The options being updated.
+	 */
+	public function intercept_currency_symbol_change( $options  ) {
+		return tribe( Base_Commerce\Settings::class )->set_default_format_on_currency_change( $options );
```
This will make us opinionated and setting what we think is better for a currency. We should allow users to use whatever they wish for the decimal and thousands separator.

rafsuntaskin replied:
It is actually just intercepting whenever you switch between currencies to avoid any mistakes made by the user. They can always use custom formats as needed. Here is a screencast detailing the issue: https://d.pr/v/yz3Pt3

juanfra replied:
Thanks for the video @rafsuntaskin. Right, that's what I'm saying. Imagine I install the plugin and go to set up Tickets Commerce. I realize the default value is USD, so I change it to Euro and I put 4 for the number of decimals, and `|` for the decimal separator. Once I hit save, those things won't be saved. 

The values that are going to be saved are the defaults, that we as a plugin in an opinionated way, decided to save. I understand where this is coming from, but I don't feel that we as a plugin should be changing the value (not even for the first time) of what a user is trying to set up for their plugin settings.

rafsuntaskin replied:
@juanfra yes, your point is valid also. That is why we will need the format preview so that customers can exactly see the reflection.

rafsuntaskin replied:
I have removed them now.

juanfra commented on src/Tickets/Commerce/Settings.php:
```diff
@@ -53,6 +53,33 @@ class Settings {
+	 */
+	public static $option_currency_thousands_separator = 'tickets-commerce-currency-thousands-separator';
+
+	/**
+	 * The option key for currency precision.
+	 *
+	 * @since TBD
+	 *
+	 * @var string
+	 */
+	public static $option_currency_precision = 'tickets-commerce-currency-precision';
```
The setting is for `Number of decimals`

juanfra commented on src/Tickets/Commerce/Settings.php:
```diff
@@ -292,6 +321,31 @@ public function get_settings() {
+
+			static::$option_currency_precision              => [
+				'type'            => 'text',
+				'label'           => esc_html__( 'Precision', 'event-tickets' ),
+				'tooltip'         => esc_html__( 'This sets the number of decimal points shown in displayed prices.', 'event-tickets' ),
```
The setting is for `Number of decimals`

juanfra commented on src/Tickets/Commerce/Utils/Currency.php:
```diff
@@ -169,6 +169,8 @@ public static function get_currency_precision( $code ) {
 			$precision = $map[ $code ]['decimal_precision'];
 		}
 
+		$precision = tribe_get_option( Settings::$option_currency_precision, $precision );
```
We should use `$number_of_decimals`, precision is confusing.

juanfra commented on src/template-tags/commerce.php:
```diff
@@ -47,3 +47,36 @@ function tec_tickets_commerce_currency_symbol() {
+ * @return int Number of decimals for the Tickets Commerce currency.
+ */
+function tec_tickets_commerce_currency_decimal_precision() {
+	return Currency::get_currency_precision( tec_tickets_commerce_currency_code() );
```
we should name everything that's related to "Number of decimals" as such.

bordoni reviewed (changes requested):
As we talked earlier today, lets make sure we add some tests for `set_default_format_on_currency_change` and how changing setting should modify the `tec_tickets` methods added here.

rafsuntaskin commented:
> As we talked earlier today, lets make sure we add some tests for `set_default_format_on_currency_change` and how changing setting should modify the `tec_tickets` methods added here.

 `set_default_format_on_currency_change` is now removed as per discussion with @juanfra . I have added tests to takr care of setting changes.

bordoni commented on src/Tickets/Commerce/Settings.php:
```diff
@@ -292,6 +319,31 @@ public function get_settings() {
+				'type'            => 'text',
+				'label'           => esc_html__( 'Decimal Separator', 'event-tickets' ),
+				'tooltip'         => esc_html__( 'This sets the decimal separator of displayed prices.', 'event-tickets' ),
+				'default'         => '.',
```
Use the default using the static variables from `Tickets/Commerce/Utils/Currency`

bordoni commented on src/Tickets/Commerce/Settings.php:
```diff
@@ -292,6 +319,31 @@ public function get_settings() {
+				'type'            => 'text',
+				'label'           => esc_html__( 'Thousands Separator', 'event-tickets' ),
+				'tooltip'         => esc_html__( 'This sets the thousand separator of displayed prices', 'event-tickets' ),
+				'default'         => ',',
```
Use the default using the static variables from `Tickets/Commerce/Utils/Currency`

bordoni commented on src/Tickets/Commerce/Settings.php:
```diff
@@ -292,6 +319,31 @@ public function get_settings() {
+				'type'            => 'text',
+				'label'           => esc_html__( 'Number of Decimals', 'event-tickets' ),
+				'tooltip'         => esc_html__( 'This sets the number of decimal points shown in displayed prices.', 'event-tickets' ),
+				'default'         => '2',
```
Use the default using the static variables from `Tickets/Commerce/Utils/Currency`

bordoni reviewed (changes requested):
Small change, but we are looking good.
